### ISSUE TITLE
OpenBLAS-0.3.11 fails to compile on ppc64le

### ISSUE THREAD

**Issue by susilehtola** [xianyi/OpenBLAS#2911]
Updating the Fedora package of OpenBLAS to version 0.3.11 fails, since the package does not build on ppc64le. I'm getting a bunch of errors like
```
../kernel/power/sgemm_kernel_power10.c:131:11: error: '__builtin_mma_xvf32gerpp' requires the '-mmma' option
  131 |           __builtin_mma_xvf32gerpp (&acc2, rowB[i], rowA[j+1]); \
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
which means you should probably add this option to the makefiles.

Also, I noticed a bunch of macro redefinitions in the power8 kernels, e.g.
```
../kernel/power/dtrmm_kernel_16x4_power8.S:85: note: this is the location of the previous definition
   85 | #define STACKSIZE 320
      | 
../kernel/power/dtrsm_kernel_LT_16x4_power8.S:51: warning: "STACKSIZE" redefined
   51 | #define STACKSIZE 520
      | 
../kernel/power/dtrsm_kernel_LT_16x4_power8.S:50: note: this is the location of the previous definition
   50 | #define STACKSIZE 320
      | 
```
which you may want to address.

**Comment by susilehtola** [comment]
Full build log
[build.log.gz](https://github.com/xianyi/OpenBLAS/files/5397873/build.log.gz)

**Comment by martin-frbg** [comment]
Thanks, looks like 0.3.11 is going to set a new record for short-livedness. At least this mma error does not appear to be my fault (the redefinitions are, I'd left the old values in when I expanded the stack to save&restore additional registers). Good to know you have access to POWER10 though...

**Comment by martin-frbg** [comment]
Ah, that is "only" a DYNAMIC_ARCH build on POWER8 ? Guess we may need to add some core-specific compiler options like it is already necessary for AVX512 on x86_64.

**Comment by martin-frbg** [comment]
The POWER10 support is only included when GCC >10.2 , and unfortunately I cannot test my current PR on Fedora32 in the OpenPOWER cloud as even with gcc 10.2.1 and the latest binutils (2.34-5.fc32) I hit an assembler error `invalid switch -mpower10`
@RajalakshmiSR do we need to (test for) a specific `as` or `binutils` version for POWER10 as well ?

**Comment by martin-frbg** [comment]
Answering my own question it appears binutils 2.35 would be required

**Comment by martin-frbg** [comment]
Hmm, do not get the -mmma error on Debian11 ppc64le (which has gcc10.2 & binutils 2.35)

**Comment by martin-frbg** [comment]
Suspect something in the way your patches modify the build options overrides or prevents the automatic addition of `-mcpu=power10 -mtune=power10 -mvsx` here https://github.com/xianyi/OpenBLAS/blob/5e8ddc9001d705473e106cf7ff285d170b7e8944/Makefile.power#L13

**Comment by RajalakshmiSR** [comment]
checking

**Comment by RajalakshmiSR** [comment]
We need gcc10.2 or gcc11 and binutils 2.35 to compile power10 files.

**Comment by martin-frbg** [comment]
May need to add a version test for `as` then (not that I like the idea, but that is what anyone on fedora32 would probably encounter right now). Main problem seems to be that his build options are somehow removing/preventing the addition of the `-mcpu=power10 -mvsx`, but perhaps that line in Makefile.power just needs to be `override COMMON_OPT +=...` for good measure.

**Comment by RajalakshmiSR** [comment]
So are you suggesting to check for binutils version before we add POWER10 in the DYNAMIC_ARCH list in Makefile.system?

**Comment by martin-frbg** [comment]
Yes, thinking about that at least, as we cannot be sure that gcc and binutils are updated together.

**Comment by RajalakshmiSR** [comment]
I will create a PR for this.

**Comment by RajalakshmiSR** [comment]
@susilehtola Is it possible to apply this patch on your fedora system and compile/check if it helps. 

diff --git a/Makefile.system b/Makefile.system
index 461f7370..44a9bbf5 100644
--- a/Makefile.system
+++ b/Makefile.system
@@ -648,12 +648,17 @@ DYNAMIC_CORE += POWER9
 else
 $(info, OpenBLAS: Your gcc version is too old to build the POWER9 kernels.)
 endif
+LDVERSIONGTEQ35  := $(shell expr `ld --version | head -1 | cut -f2 -d "."  | cut -f1 -d "-"` \>= 35)
 ifeq ($(GCCVERSIONGTEQ11), 1)
+ifeq ($(LDVERSIONGTEQ35), 1)
 DYNAMIC_CORE += POWER10
+endif
 else ifeq ($(GCCVERSIONGTEQ10), 1)
 ifeq ($(GCCMINORVERSIONGTEQ2), 1)
+ifeq ($(LDVERSIONGTEQ35), 1)
 DYNAMIC_CORE += POWER10
 endif
+endif
 else
 $(info, OpenBLAS: Your gcc version is too old to build the POWER10 kernels.)
 endif

**Comment by RajalakshmiSR** [comment]
@susilehtola Can you also confirm your gcc and binutils version ? Can you paste full output of gcc --version and ld --version ?

**Comment by RajalakshmiSR** [comment]
Looking at the log, it is clear that -mcpu=power10 is not passed here. So the actual error is not due to binutils/gcc versions 

 .
gcc -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -mcpu=power8 -mtune=power8 -fasynchronous-unwind-tables -fstack-clash-protection -fPIC -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DDYNAMIC_ARCH -DDYNAMIC_OLDER -DNO_WARMUP -DMAX_CPU_NUMBER=128 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.11\" -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME=sgemm_kernel_POWER10 -DASMFNAME=sgemm_kernel_POWER10_ -DNAME=sgemm_kernel_POWER10_ -DCNAME=sgemm_kernel_POWER10 -DCHAR_NAME=\"sgemm_kernel_POWER10_\" -DCHAR_CNAME=\"sgemm_kernel_POWER10\" -DNO_AFFINITY -DTS=_POWER10 -I.. -DBUILD_KERNEL -DTABLE_NAME=gotoblas_POWER10 -UDOUBLE  -UCOMPLEX -c -UDOUBLE -UCOMPLEX ../kernel/power/sgemm_kernel_power10.c -o sgemm_kernel_POWER10.o

**Comment by susilehtola** [comment]
@RajalakshmiSR please post a proper patch, that one is broken.

Also, the flags are determined by Fedora policy as
```
COMMON="%{optflags} -fPIC"
FCOMMON="$COMMON -frecursive"
make -C serial     $TARGET USE_THREAD=0 USE_LOCKING=1 USE_OPENMP=0 FC=gfortran CC=gcc COMMON_OPT="$COMMON" FCOMMON_OPT="$FCOMMON" $NMAX LIBPREFIX="libopenblas"      $AVX $LAPACKE INTERFACE64=0
```
If some object files need special mcpu flags, they should be specified in the respective make rule.

**Comment by RajalakshmiSR** [comment]
> Suspect something in the way your patches modify the build options overrides or prevents the automatic addition of `-mcpu=power10 -mtune=power10 -mvsx` here
> 
> https://github.com/xianyi/OpenBLAS/blob/5e8ddc9001d705473e106cf7ff285d170b7e8944/Makefile.power#L13

@martin-frbg The only difference in passing the flags between power9 and power10 in Makefile.power is Power9 is using  CCOMMON_OPT whereas power10 uses COMMON_OPT.

**Comment by martin-frbg** [comment]
@susielethola can you try changing that line in Makefile.power that we both quoted above so that it starts with `override` before the CCOMMON_OPT ?

**Comment by RajalakshmiSR** [comment]
@susilehtola can you try changing COMMON_OPT to CCOMMON_OPT in OpenBLAS/Makefile.power line 13

**Comment by martin-frbg** [comment]
@RajalakshmiSR of course you are right, I seem to have overlooked the missing "C". Updated my PR #2911 accordingly but have not added the binutils version check in Makefile.system yet.

**Comment by susilehtola** [comment]
@martin-frbg adding the override fixes that issue, but now I see there's a linking problem, see #2922

**Comment by susilehtola** [comment]
@RajalakshmiSR new build running

**Comment by martin-frbg** [comment]
@susilehtola thanks, but actually Rajalakshmi spotted the actual problem and proposed the correct solution hours before I wrote mine without properly reading. Giving that variable an extra C is the way to go. 
The link errors are another sad saga that has seen several partial fixes by now - in short I had introduced multiple typos and logic errors in the gensymbol script when I rearranged functions by argument type, and there was nothing missing (i.e. a missing blank in a crucial spot) in its invocation from the Makefile as well. #2909 and #2919 fix most of it but I am still fixing more.

**Comment by RajalakshmiSR** [comment]
> @RajalakshmiSR of course you are right, I seem to have overlooked the missing "C". Updated my PR #2911 accordingly but have not added the binutils version check in Makefile.system yet.

@martin-frbg Is it possible to try (manually edit Makefile.system) this https://github.com/xianyi/OpenBLAS/issues/2911#issuecomment-711439168 in your openpower cloud system (fedora32) which has new gcc and older binutils..  (to check if it fixes that invalid switch error in https://github.com/xianyi/OpenBLAS/issues/2911#issuecomment-711412747).

**Comment by RajalakshmiSR** [comment]
> Yes, thinking about that at least, as we cannot be sure that gcc and binutils are updated together.

This is a generic problem and I think we don't have to fix this in a hurry only for POWER10 for 0.3.11 release. We can create a separate issue and discuss this.

**Comment by martin-frbg** [comment]
Certainly, and I expect it will fix the problem.

**Comment by martin-frbg** [comment]
Meh, this leads to a new problem later as dynamic_power.c does not "know" that gotoblas_POWER10 did not get built (as it only has ifdefs for `__GNUC__`). I guess the tests for defining HAVE_P10_SUPPORT should be moved into c_check or getarch (or as a quick and dirty fix, that spot in Makefile.power should also add -DHAVE_P10_SUPPORT to CCOMMON_OPT)

**Comment by RajalakshmiSR** [comment]
@martin-frbg Yes.. (expected)... this needs more testing with different combinations of compiler versions. I can fix this post release as I do not want to introduce new bugs by adding this.

**Comment by martin-frbg** [comment]
I'll test moving the `define HAVE_P10_SUPPORT` from dynamic_power.c to Makefile.power for now - this would at least be consistent and should not hurt non-DYNAMIC_ARCH builds (famous last words)

**Comment by RajalakshmiSR** [comment]
With the current master,
1. Non dynamic arch builds are not affected. 
2. DYNAMIC_ARCH is affected only when binutils is old and gcc is new.

**Comment by martin-frbg** [comment]
Yes, I'm only worried about (2) because that seems to be the current state of the ppc64le Fedora. Of course we can leave it at that for now, I certainly do not want two broken releases in a row.

**Comment by RajalakshmiSR** [comment]
@martin-frbg can you update binutils on your fedora32. ?

**Comment by martin-frbg** [comment]
I could certainly build from source, the only issue is that there appears ot be no binary update package available (yet ?) via their normal update mechanism.

**Comment by RajalakshmiSR** [comment]
> I'll test moving the `define HAVE_P10_SUPPORT` from dynamic_power.c to Makefile.power for now - this would at least be consistent and should not hurt non-DYNAMIC_ARCH builds (famous last words)
@martin-frbg 
Are you suggesting this for now? 
[patch.txt](https://github.com/xianyi/OpenBLAS/files/5405099/patch.txt)

**Comment by martin-frbg** [comment]
Yes, with an additional "define HAVE_P10_SUPPORT" a few lines above, in the `ifneq ($(C_COMPILER), GCC)` block so that it matches the previous (intended) behaviour for other compilers as well. Would you agree to this ?

**Comment by RajalakshmiSR** [comment]
Yes.

**Comment by martin-frbg** [comment]
Thx. This is #2925 now (though I notice we'll need a proper check in the cmake part eventually)

**Comment by susilehtola** [comment]
> @susilehtola can you try changing COMMON_OPT to CCOMMON_OPT in OpenBLAS/Makefile.power line 13

Doesn't this still need to be done though?

Will you be releasing 0.3.12 soon or should I backport the changes to 0.3.11?

**Comment by martin-frbg** [comment]
> Doesn't this still need to be done though?

That was the "fix power10 option passing" bit of #2916 
I do not intend to sit on the steaming pile of ... that was 0.3.11 for longer than I have to, I am just currently trying to convince myself that what we have now has no glaring bugs left. I was prepared for a bit of fallout from the big makefile reshuffle, but I certainly did not expect anything like this. And besides hating myself for rushing the release after becoming target-fixated with getting all Travis checks passing again, I am a bit worried that none of the current Ci tests gave an unmistakable warning.
